src/Microsoft.Data.SqlClient/tests/ManualTests/AlwaysEncrypted/ApiShould.cs

```diff
@@ -2114,6 +2117,93 @@ public void TestSqlCommandCancellationToken(string connection, int initalValue,
             }
         }
 
+        [ConditionalTheory(typeof(DataTestUtility), nameof(DataTestUtility.AreConnStringSetupForAE))]
+        [ClassData(typeof(AEConnectionStringProvider))]
+        public void TestCustomKeyStoreProviderRegistration(string connectionString)
+        {
+            if (!SQLSetupStrategyAzureKeyVault.isAKVProviderRegistered)
+            {
+                SqlColumnEncryptionAzureKeyVaultProvider sqlColumnEncryptionAzureKeyVaultProvider =
+                    new SqlColumnEncryptionAzureKeyVaultProvider(new SqlClientCustomTokenCredential());
+
+                DummyKeyStoreProvider dummyProvider = new DummyKeyStoreProvider();
+
+                SqlConnection.RegisterColumnEncryptionKeyStoreProviders(customProviders:
+                    new Dictionary<string, SqlColumnEncryptionKeyStoreProvider>(capacity: 2, comparer: StringComparer.OrdinalIgnoreCase)
+                    {
+                        { "AZURE_KEY_VAULT", sqlColumnEncryptionAzureKeyVaultProvider},
+                        { "DummyProvider", dummyProvider}
+                    });
+
+                SQLSetupStrategyAzureKeyVault.isAKVProviderRegistered = true;
+            }
+
+            Dictionary<string, SqlColumnEncryptionKeyStoreProvider> requiredProvider =
+                new Dictionary<string, SqlColumnEncryptionKeyStoreProvider>()
+                {
+                    { "DummyProvider", new DummyKeyStoreProvider() }
+                };
+
+            Dictionary<string, SqlColumnEncryptionKeyStoreProvider> notRequiredProvider =
+                new Dictionary<string, SqlColumnEncryptionKeyStoreProvider>()
+                {
+                    { "DummyProvider2", new DummyKeyStoreProvider() }
+                };
+
+            string failedToDecryptErrorMessage = "Failed to decrypt a column encryption key using key store provider: " +
+                "'DummyProvider'. Verify the properties of the column encryption key and its column master key in your database.";
+            string providerNotImplementedMessage = "The method or operation is not implemented.";
+            string providerNotFoundMessage = "Invalid key store provider name: 'DummyProvider'. A key store " +
+                "provider name must denote either a system key store provider or a registered custom key store provider." +
+                " Valid system key store provider names are: 'MSSQL_CERTIFICATE_STORE', 'MSSQL_CNG_STORE', " +
+                "'MSSQL_CSP_PROVIDER'. Valid (currently registered) custom key store provider names are: 'DummyProvider2'.";
```

Comment:
You can use `SystemDataResourceManager` class to resolve the messages from the string repository. It'll easier to maintain except support multi-culture.